storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentReader.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2017-present Open Networking Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.atomix.storage.journal;
+
+import io.atomix.storage.StorageException;
+import io.atomix.storage.journal.index.JournalIndex;
+import io.atomix.storage.journal.index.Position;
+import io.atomix.utils.serializer.Namespace;
+
+import java.io.IOException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.NoSuchElementException;
+import java.util.zip.CRC32;
+import java.util.zip.Checksum;
+
+/**
+ * Log segment reader.
+ *
+ * @author <a href="http://github.com/kuujo">Jordan Halterman</a>
+ */
+class FileChannelJournalSegmentReader<E> implements JournalReader<E> {
+  private final FileChannel channel;
+  private final int maxEntrySize;
+  private final JournalIndex index;
+  private final Namespace namespace;
+  private final ByteBuffer memory;
+  private final long firstIndex;
+  private Indexed<E> currentEntry;
+  private Indexed<E> nextEntry;
+
+  FileChannelJournalSegmentReader(
+      FileChannel channel,
+      JournalSegment<E> segment,
+      int maxEntrySize,
+      JournalIndex index,
+      Namespace namespace) {
+    this.channel = channel;
+    this.maxEntrySize = maxEntrySize;
+    this.index = index;
+    this.namespace = namespace;
+    this.memory = ByteBuffer.allocate((maxEntrySize + Integer.BYTES + Integer.BYTES) * 2);
+    this.firstIndex = segment.index();
+    reset();
+  }
+
+  @Override
+  public long getFirstIndex() {
+    return firstIndex;
+  }
+
+  @Override
+  public long getCurrentIndex() {
+    return currentEntry != null ? currentEntry.index() : 0;
+  }
+
+  @Override
+  public Indexed<E> getCurrentEntry() {
+    return currentEntry;
+  }
+
+  @Override
+  public long getNextIndex() {
+    return currentEntry != null ? currentEntry.index() + 1 : firstIndex;
```

Comment:
are entries guaranteed to be contiguous? why aren't we using nextEntry here?

Reply:
Yes they're guaranteed to be contiguous. `nextEntry` gets its index from `getNextIndex()`, so I think there might be some weird problems with using it.